pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,144 @@
+	initialAccumulator DebouncedEvent
+
+	running bool
+}
+
+// NewDebouncedChannel create a new instance of DebouncedChannel
+// debounceAfter: A new outgoing event is emitted if no additional event is enqueued during this period (resets on every enqueued event).
+// debounceMax: A new outgoing event is emitted after this period (even if additional events were enqueued in that period).
+// initialAccumulator: Initial value of the accumulator
+// merge: function, which is called to merge two events if they are debounced.
+func NewDebouncedChannel(debounceAfter time.Duration, debounceMax time.Duration,
+	initialAccumulator DebouncedEvent, merge func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent) *DebouncedChannel {
+
+	c := make(chan func(func(DebouncedEvent)), 1)
+
+	return &DebouncedChannel{
+		C:                  c,
+		c:                  c,
+		initialAccumulator: initialAccumulator,
+		accumulator:        initialAccumulator,
+		merge:              merge,
+		debounceAfter:      debounceAfter,
+		debounceMax:        debounceMax,
+		lastEmitTime:       time.Unix(0, 0),
+	}
+
+}
+
+// Enqueue pushes a new event to the channel
+func (dc *DebouncedChannel) Enqueue(element DebouncedEvent) {
+	dc.mutex.Lock()
+	defer dc.mutex.Unlock()
+	dc.accumulator = dc.merge(dc.accumulator, element)
+	if dc.debounceCounter == 0 {
+		dc.debounceMaxEnd = time.Now().Add(dc.debounceMax)
+		go func() {
+			dc.mutex.Lock()
+			defer dc.mutex.Unlock()
+
+			for {
+				debounceCounter := dc.debounceCounter
+
+				sleepDuration := min(dc.debounceAfter, time.Until(dc.debounceMaxEnd))
+				dc.mutex.Unlock()
+				time.Sleep(sleepDuration)
+				dc.mutex.Lock()
+
+				if debounceCounter == dc.debounceCounter || time.Now().After(dc.debounceMaxEnd) {
+					break
+				}
+			}
+			dc.runIfNeeded()
+		}()
+	}
+
+	dc.debounceAfterEnd = time.Now().Add(dc.debounceAfter)
+	dc.debounceCounter++
+
+	dc.lastEnqueueTime = time.Now()
+
+	dc.runIfNeeded()
+}
+
+func min(a, b time.Duration) time.Duration {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (dc *DebouncedChannel) runIfNeeded() {
+	if dc.debounceCounter > 0 && !dc.running && (dc.debounceMaxEnd.Before(time.Now()) || dc.debounceAfterEnd.Before(time.Now())) {
+
+		accumulator := dc.accumulator
+		debounceCounter := dc.debounceCounter
+		pushCounter := dc.pushCounter
+		lastEnqueueTime := dc.lastEnqueueTime
+		lastEmitTime := dc.lastEmitTime
+
+		dc.running = true
+		dc.debounceCounter = 0
+		dc.pushCounter++
+		dc.accumulator = dc.initialAccumulator
+		dc.lastEmitTime = time.Now()
+
+		dc.c <- func(callback func(DebouncedEvent)) {
+			adsLog.Infof("Push debounce stable[%d] %d: %v since last change, %v since last push, accumulator=%v",
+				pushCounter, debounceCounter, time.Since(lastEnqueueTime), lastEmitTime, accumulator)
+
+			callback(accumulator)
+			dc.done()
+		}
+	}
+}
+
+func (dc *DebouncedChannel) done() {
+	dc.mutex.Lock()
+	defer dc.mutex.Unlock()
+	dc.running = false
+	dc.runIfNeeded()
```

Comment:
I am not sure why this function is called many places: runIfNeeded->dc.done()->runIfNeeded

Reply:
Maybe the function should be named `runIfNeededAndPossible`
The function checks for several conditions and those needs to be called whenever one (or more) of this conditions change.
The conditions are:
1. at least one of the debounce timers has run out
1. there is currently no job running
1. there is an pending event

For example, it is called in the `done` function, because condition 2 has changed and we need to recheck condition 1 and 3 (and kick-of a new event if enqueue was called while the job was running and debouncing has finished).

Reply:
The sequence `runIfNeeded->dc.done()->runIfNeeded` is not correct. `done` is called from the receiver to signal that the event is processed.

```
         DebouncedChannel                              Channel                  Receiver
              | --+                                      |                          |---+
              |   | runIf Needed                         |                          |---+
              |   +------ write to cannel .   ----->     |    <--- receive ---      |
              |   |                                      |                          |---+
- Enqueue ->  |<--+                                      |                          |   |
- Enqueue ->  |                                          |                          |  process
              |   <---------------------------------done ---------------------------|<--+```